### PR TITLE
feat(android): add v8 heap tool

### PR DESCRIPTION
**Pull request by linjinze999**
Before submitting a new pull request, please make sure:

- [x] Test cases have been added/updated/passed for the code you will submit.
- [x] Documentation has added or updated.
- [x] Commit message is following the [Convention Commit](https://conventionalcommits.org/) guideline with maximum 72 characters.
- [x] Squash the repeat code commits, short patches are welcome.

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/Hippy?pullRequest=1803) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/Hippy?pullRequest=1803) before we can accept your contribution.<br/><hr/>**jinzelin** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/Hippy?pullRequest=1803) it.</sub>

File: android/sdk/src/main/jni/include/bridge/entry.h [hunk 1]
```diff
@@ -54,5 +68,84 @@ jboolean RunScriptFromUri(JNIEnv* j_env,
...
+        save_res_ = sOk;
+        return;
+    }
+    int GetChunkSize(){
+        return 1024;
+    }
+    WriteResult WriteAsciiChunk(char* data, int size) {
+        if (!base::StringViewUtils::IsEmpty(m_file_path)) {
+            // save to file
+            std::string heap_snapshot_chunk_data =
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/include/bridge/entry.h:108]: 推荐带上size，另外这里为什么要转u8呢，data是buffer还是string？

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 2]
```diff
@@ -600,6 +618,148 @@ void DestroyInstance(__unused JNIEnv* j_env,
...
 }
 
+// [Heap] GetHeapCodeStatistics
+jobject GetHeapCodeStatistics(__unused JNIEnv* j_env,
+                             __unused jobject j_object,
+                             jlong j_runtime_id) {
+  TDF_BASE_DLOG(INFO) << "GetHeapCodeStatistics begin, j_runtime_id = " << j_runtime_id;
+  auto runtime_id = static_cast<int32_t>(j_runtime_id);
+  std::shared_ptr<Runtime> runtime = Runtime::Find(runtime_id);
+  v8::HeapCodeStatistics* heap_code_statistics = new v8::HeapCodeStatistics();
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/src/bridge/entry.cc:628]: 最好用智能指针，不要用显示的new，需要用指针的地方用get获取裸指针。因为后续如果出现异常，极其容易遗忘delete

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 3]
```diff
@@ -600,6 +618,148 @@ void DestroyInstance(__unused JNIEnv* j_env,
...
+  if (runtime) {
+      v8::Isolate* isolate = std::static_pointer_cast<hippy::napi::V8VM>(runtime->GetEngine()->GetVM())->isolate_;
+      isolate->GetHeapCodeAndMetadataStatistics(heap_code_statistics);
+  } else {
+      TDF_BASE_DLOG(WARNING) << "GetHeapCodeStatistics, j_runtime_id invalid";
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/src/bridge/entry.cc:633]: 出现异常了，理论应该需要返回错误，提前返回。后续已经没必要执行了

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 4]
```diff
@@ -600,6 +618,148 @@ void DestroyInstance(__unused JNIEnv* j_env,
...
+  }
+  // set HeapCodeStatistics data
+  jclass java_class_hcs = j_env->FindClass("com/tencent/mtt/hippy/bridge/heap/HeapCodeStatistics");
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/src/bridge/entry.cc:636]: 是否遗忘了DeleteLocalRef

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 5]
```diff
@@ -600,6 +618,148 @@ void DestroyInstance(__unused JNIEnv* j_env,
...
+  delete heap_code_statistics;
+  heap_code_statistics = nullptr;
+  TDF_BASE_DLOG(INFO) << "GetHeapCodeStatistics end";
+  return hcs_obj;
+}
+// [Heap] GetHeapSpaceStatisticsList
+jobject GetHeapSpaceStatisticsList(__unused JNIEnv* j_env,
+                                   __unused jobject j_object,
+                                   jlong j_runtime_id) {
+    TDF_BASE_DLOG(INFO) << "GetHeapSpaceStatisticsList begin, j_runtime_id = " << j_runtime_id;
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/src/bridge/entry.cc:652]: j_runtime_id 异常时及早返回，也就没必要各种反射查找jclass了

Comment (churchill-zhang) [reply android/sdk/src/main/jni/src/bridge/entry.cc:652]: 另外jni相关对象最好j_xxx命名，方便大家理解

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 6]
```diff
@@ -600,6 +618,148 @@ void DestroyInstance(__unused JNIEnv* j_env,
...
+            jobject hss_obj = j_env->NewObject(java_class_hss,
+                                               hss_constructor,
+                                               j_env->NewStringUTF(heap_space_statistics->space_name()),
+                                               (long)heap_space_statistics->space_size(),
+                                               (long)heap_space_statistics->space_used_size(),
+                                               (long)heap_space_statistics->space_available_size(),
+                                               (long)heap_space_statistics->physical_space_size());
+            j_env->CallBooleanMethod(list_obj, list_add, hss_obj);
+            delete heap_space_statistics;
+            heap_space_statistics = nullptr;
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/src/bridge/entry.cc:682]: 非返回值的jobject需要显式释放引用

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 7]
```diff
@@ -600,6 +618,148 @@ void DestroyInstance(__unused JNIEnv* j_env,
...
+                          jlong j_runtime_id) {
+    TDF_BASE_DLOG(INFO) << "GetHeapStatistics begin, j_runtime_id = " << j_runtime_id;
+    v8::HeapStatistics* heap_statistics = new v8::HeapStatistics();
+    auto runtime_id = static_cast<int32_t>(j_runtime_id);
+    std::shared_ptr<Runtime> runtime = Runtime::Find(runtime_id);
+    if (runtime) {
+        v8::Isolate* isolate = std::static_pointer_cast<hippy::napi::V8VM>(runtime->GetEngine()->GetVM())->isolate_;
+        isolate->GetHeapStatistics(heap_statistics);
+    } else {
+        TDF_BASE_DLOG(WARNING) << "GetHeapStatistics, j_runtime_id invalid";
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/src/bridge/entry.cc:702]: 同上，及早并释放资源

File: android/sdk/src/main/jni/include/bridge/entry.h [hunk 8]
```diff
@@ -22,10 +22,24 @@
...
+#include <core/base/file.h>
 #include <jni.h>
+#include <sys/stat.h>
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wsign-conversion"
+#include "v8/v8-profiler.h"
+#pragma clang diagnostic pop
+#include "jni/jni_utils.h"
+
+#define HSOSA_FILE_OK 0
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/include/bridge/entry.h:34]: c++推荐用枚举

File: android/sdk/src/main/jni/src/bridge/entry.cc [hunk 9]
```diff
@@ -600,6 +618,148 @@ void DestroyInstance(__unused JNIEnv* j_env,
...
+// [Heap] GetHeapCodeStatistics
+jobject GetHeapCodeStatistics(__unused JNIEnv* j_env,
+                             __unused jobject j_object,
+                             jlong j_runtime_id) {
+  TDF_BASE_DLOG(INFO) << "GetHeapCodeStatistics begin, j_runtime_id = " << j_runtime_id;
+  auto runtime_id = static_cast<int32_t>(j_runtime_id);
+  std::shared_ptr<Runtime> runtime = Runtime::Find(runtime_id);
+  v8::HeapCodeStatistics* heap_code_statistics = new v8::HeapCodeStatistics();
+  if (runtime) {
+      v8::Isolate* isolate = std::static_pointer_cast<hippy::napi::V8VM>(runtime->GetEngine()->GetVM())->isolate_;
```
Comment (churchill-zhang) [inline android/sdk/src/main/jni/src/bridge/entry.cc:630]: 另外v8相关操作需要在v8线程，否则会crash。这里看不出来java一定是在js线程调用，那就需要c++显示postTask到js线程执行

**Review by medns — CHANGES_REQUESTED** [review]
Request changes

**Comment by linjinze999** [comment]
先关闭，另外提一个Pr，减少commit

**Pull request closed without merging**